src/Compilers/CSharp/Portable/Syntax/CSharpSyntaxNode.cs

```diff
@@ -91,21 +91,25 @@ private static void ComputeSyntaxTree(CSharpSyntaxNode node)
                 var parent = node.Parent;
                 if (parent == null)
                 {
-                    Interlocked.CompareExchange(ref node._syntaxTree, CSharpSyntaxTree.CreateWithoutClone(node), null);
-                    tree = node._syntaxTree;
+                    // set the tree on the root node atomically
+                    tree = CSharpSyntaxTree.CreateWithoutClone(node);
+                    var orig = Interlocked.CompareExchange(ref node._syntaxTree, tree, null);
```

Comment:
That would work too.
I expect that this Exch will be typically successful so predictably I will not need to re-read that field.